File: src/services/appointmentService.ts
import { Appointment, AppointmentSlot, AppointmentFilters } from '../types/appointment';

// API service for appointment management
export const appointmentService = {
  // Get all appointment slots
  async getSlots(): Promise<AppointmentSlot[]> {
    try {
      const response = await fetch('/api/admin/slots');
      if (!response.ok) throw new Error('Failed to fetch slots');
      return await response.json();
    } catch (error) {
      console.error('Error fetching appointment slots:', error);
      throw error;
    }
  },

  // Create a new appointment slot
  async createSlot(slot: Omit<AppointmentSlot, 'id'>): Promise<AppointmentSlot> {
    try {
      const response = await fetch('/api/admin/slots', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(slot),
      });
      if (!response.ok) throw new Error('Failed to create slot');
      return await response.json();
    } catch (error) {
      console.error('Error creating appointment slot:', error);
      throw error;
    }
  },

  // Update an existing appointment slot
  async updateSlot(id: string, slot: Partial<AppointmentSlot>): Promise<AppointmentSlot> {
    try {
      const response = await fetch(`/api/admin/slots/${id}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(slot),
      });
      if (!response.ok) throw new Error('Failed to update slot');
      return await response.json();
    } catch (error) {
      console.error('Error updating appointment slot:', error);
      throw error;
    }
  },

  // Delete an appointment slot
  async deleteSlot(id: string): Promise<void> {
    try {
      const response = await fetch(`/api/admin/slots/${id}`, {
        method: 'DELETE',
      });
      if (!response.ok) throw new Error('Failed to delete slot');
    } catch (error) {
      console.error('Error deleting appointment slot:', error);
      throw error;
    }
  },

  // Get all appointments with optional filters
  async getAppointments(filters?: AppointmentFilters): Promise<Appointment[]> {
    try {
      let url = '/api/admin/appointments';
      if (filters) {
        const params = new URLSearchParams();
        if (filters.startDate) params.append('startDate', filters.startDate);
        if (filters.endDate) params.append('endDate', filters.endDate);
        if (filters.status) params.append('status', filters.status);
        if (filters.serviceId) params.append('serviceId', filters.serviceId);
        url += `?${params.toString()}`;
      }

      const response = await fetch(url);
      if (!response.ok) throw new Error('Failed to fetch appointments');
      return await response.json();
    } catch (error) {
      console.error('Error fetching appointments:', error);
      throw error;
    }
  },

  // Update appointment status
  async updateAppointmentStatus(id: string, status: 'scheduled' | 'completed' | 'cancelled'): Promise<Appointment> {
    try {
      const response = await fetch(`/api/admin/appointments/${id}/status`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ status }),
      });
      if (!response.ok) throw new Error('Failed to update appointment status');
      return await response.json();
    } catch (error) {
      console.error('Error updating appointment status:', error);
      throw error;
    }
  },
};

File: app/services/appointment_service.py
from datetime import datetime, timedelta
from typing import List, Optional
from fastapi import HTTPException, status
from uuid import UUID

from app.db.repositories.appointment_repository import AppointmentRepository
from app.models.appointment import Appointment, AppointmentStatus
from app.schemas.appointment import (
    AppointmentCreate,
    AppointmentUpdate,
    AppointmentSlotCreate,
    AppointmentSlot,
    AppointmentAdminResponse
)

class AppointmentService:
    def __init__(self, appointment_repository: AppointmentRepository):
        self.repository = appointment_repository

    async def create_appointment_slot(self, slot_data: AppointmentSlotCreate) -> AppointmentSlot:
        """Create available appointment slots as admin"""
        # Check if slots already exist for this time
        existing_slots = await self.repository.get_slots_by_time_range(
            location_id=slot_data.location_id,
            start_time=slot_data.start_time,
            end_time=slot_data.end_time
        )

        if existing_slots:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Appointment slots already exist for this time range"
            )

        return await self.repository.create_appointment_slot(slot_data)

    async def get_all_slots(
        self,
        location_id: Optional[UUID] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        available_only: bool = False
    ) -> List[AppointmentSlot]:
        """Get all appointment slots with optional filtering"""
        if not start_date:
            start_date = datetime.now()
        if not end_date:
            end_date = start_date + timedelta(days=30)

        return await self.repository.get_appointment_slots(
            location_id=location_id,
            start_date=start_date,
            end_date=end_date,
            available_only=available_only
        )

    async def update_appointment_slot(self, slot_id: UUID, slot_data: AppointmentSlotCreate) -> AppointmentSlot:
        """Update an appointment slot as admin"""
        slot = await self.repository.get_slot_by_id(slot_id)
        if not slot:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Appointment slot not found"
            )

        # Check if the slot has booked appointments
        appointments = await self.repository.get_appointments_by_slot_id(slot_id)
        if any(appointment.status != AppointmentStatus.CANCELLED for appointment in appointments):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Cannot modify slot with active appointments"
            )

        return await self.repository.update_appointment_slot(slot_id, slot_data)

    async def delete_appointment_slot(self, slot_id: UUID) -> bool:
        """Delete an appointment slot if no active appointments"""
        slot = await self.repository.get_slot_by_id(slot_id)
        if not slot:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Appointment slot not found"
            )

        # Check if the slot has booked appointments
        appointments = await self.repository.get_appointments_by_slot_id(slot_id)
        if any(appointment.status != AppointmentStatus.CANCELLED for appointment in appointments):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Cannot delete slot with active appointments"
            )

        return await self.repository.delete_appointment_slot(slot_id)

    async def get_all_appointments(
        self,
        location_id: Optional[UUID] = None,
        status: Optional[AppointmentStatus] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        user_id: Optional[UUID] = None
    ) -> List[AppointmentAdminResponse]:
        """Get all appointments with optional filtering for admin view"""
        if not start_date:
            start_date = datetime.now() - timedelta(days=7)  # Include recent past appointments
        if not end_date:
            end_date = datetime.now() + timedelta(days=30)

        appointments = await self.repository.get_appointments(
            location_id=location_id,
            status=status,
            start_date=start_date,
            end_date=end_date,
            user_id=user_id
        )

        # Convert to admin response format with user and application details
        return [
            await self._enrich_appointment_with_details(appointment)
            for appointment in appointments
        ]

    async def _enrich_appointment_with_details(self, appointment: Appointment) -> AppointmentAdminResponse:
        """Enrich appointment with user and application details"""
        # In a real implementation, this would fetch user and application details
        # For now we'll return a simplified version
        return AppointmentAdminResponse(
            id=appointment.id,
            slot_id=appointment.slot_id,
            user_id=appointment.user_id,
            application_id=appointment.application_id,
            status=appointment.status,
            created_at=appointment.created_at,
            updated_at=appointment.updated_at,
            location_id=appointment.location_id,
            start_time=appointment.start_time,
            end_time=appointment.end_time,
            user={
                "id": appointment.user_id,
                "email": "",  # Would be populated from user service
                "first_name": "",
                "last_name": ""
            },
            application={
                "id": appointment.application_id,
                "type": "",  # Would be populated from application service
                "status": ""
            }
        )
